File: java/matrixkalq/MatrixBase.java
/* MatrixBase.java */

package matrixkalq;

//import java.lang.Cloneable;
//import java.lang.IllegalArgumentException;
import java.io.PrintStream;


//Abstract Base Class
public abstract class MatrixBase implements Cloneable {
	
	//Campi dati:
		private int rows;
		private int cols;
		
	//Costruttori
		public MatrixBase(int r, int c) {
			if(r<0 || c<0) {
				rows=0; cols=0;
				throw new IllegalArgumentException("creazione di una matrice con dimensioni negative non permessa");
			}
			rows=r; cols=c;
		}
		public MatrixBase() { this(0,0); }
		
	//Metodi virtuali puri:
		public abstract double getElem(int i, int j);
		public abstract void setElem(int i, int j, double val);
		public abstract void swapRows(int r1, int r2, int fromCol);
		public abstract void swapCols(int c1, int c2, int fromRow);
		public abstract void print(PrintStream ps, String separ);
		public abstract void load(String s);
		public abstract int rank();
		public abstract MatrixBase sum(MatrixBase m);
		public abstract MatrixBase dif(MatrixBase m);
		public abstract MatrixBase sProd(double scalar);
		public abstract MatrixBase trans();
		
	//Metodi propri:
		public int numRows() { return rows; }
		public int numCols() { return cols; }
		
		public MatrixBase clone() {
			MatrixBase aux = null;
			try {aux = (MatrixBase)super.clone(); }
			catch(CloneNotSupportedException e) { throw new RuntimeException("clone() ha fallito", e); }
			
			aux.rows = this.rows;
			aux.cols = this.cols;
			
			return aux;
		}
		
		public boolean equals(Object obj) {
			if(this == obj)
				return true;
			
			if(obj==null || getClass()!=obj.getClass())
				return false;
			
			MatrixBase other = (MatrixBase)obj;
			if(!(numRows()==other.numRows() && numCols()==other.numRows()))
				return false;
			
			for(int i=0; i<numRows(); ++i) //o: i<other.numRows()
				for(int j=0; j<numCols(); ++j) //o: j<other.numCols()
					if(getElem(i,j) != other.getElem(i,j))
						return false;
			
			return true;
		}
		
	//Eccezioni:
		//	IllegalArgumentException: MatrixBase(..)
		//	RuntimeException: clone()
	
} //fine MatrixBase

File: cplusplus/Core/MatrixBase.cpp
/* MatrixBase.cpp */
#include "MatrixBase.h"


//Metodi interni:

MatrixBase::MatrixBase(unsigned int r, unsigned int c): rows(r), cols(c) {}

MatrixBase::~MatrixBase() {}

unsigned int MatrixBase::numRows() const {
	return rows;
}

unsigned int MatrixBase::numCols() const {
	return cols;
}


//Funzioni esterne:

ostream& operator<<(ostream& os, const MatrixBase& m) {
	m.print(os);
	return os;
}

istream& operator>>(istream& is, MatrixBase& m) {
	m.load(is);
	return is;
}

bool operator==(const MatrixBase& m1, const MatrixBase& m2) {
	bool uguali = true;
	if(!(m1.numRows()==m2.numRows() && m1.numCols()==m2.numCols()))
		uguali = false;
	else
		for(unsigned int i=0; i<m1.numRows() && uguali; ++i) //o: i<m2.numRows()
			for(unsigned int j=0; j<m1.numCols() && uguali; ++j) //o: j<m2.numCols()
				if(m1.at(i,j) != m2.at(i,j))
					uguali = false;
	
	return uguali;
}

bool operator!=(const MatrixBase& m1, const MatrixBase& m2) {
	return !(m1==m2);
}